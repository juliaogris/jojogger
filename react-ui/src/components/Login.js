import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { login, signup } from '../util/api'
import { getEmailError, getPasswordError } from '../util/funcs'

export default class Login extends Component {
  constructor () {
    super()
    this.state = {
      email: '',
      password: ''
    }
    this.handleLoginClick = this.handleLoginClick.bind(this)
    this.handleSignupClick = this.handleSignupClick.bind(this)
    this.handleInputChange = this.handleInputChange.bind(this)
  }

  handleLoginClick (event) {
    event.preventDefault()
    const { email, password } = this.state
    const { setLoading, setUser, handleAuthError } = this.props
    const error = getEmailError(email) || getPasswordError(password)
    if (error) {
      handleAuthError(error)
      return
    }

    setLoading(true)
    login(email, password)
      .then(user => setUser(user))
      .catch(error => handleAuthError(error))
  }

  handleSignupClick (event) {
    event.preventDefault()
    const { setLoading, setUser, handleAuthError } = this.props
    const { email, password } = this.state
    const error = getEmailError(email) || getPasswordError(password)
    if (error) {
      handleAuthError(error)
      return
    }

    setLoading(true)
    signup(email, password)
      .then(user => setUser(user))
      .catch(error => handleAuthError(error))
  }

  handleInputChange (event) {
    const { name, value } = event.target
    this.setState({ [name]: value })
  }

  render () {
    const { email, password } = this.state
    const { authError } = this.props
    return (
      <form className='login' action={this.login}>
        {authError && <div className='auth-error'>{authError.message}</div>}
        <input
          placeholder='email@example.com'
          type='email'
          name='email'
          value={email}
          onChange={this.handleInputChange}
        />
        <input
          placeholder='your password'
          type='password'
          name='password'
          value={password}
          onChange={this.handleInputChange}
          />
        <div className='button-row'>
          <button onClick={this.handleLoginClick}> Log in </button>
          <button onClick={this.handleSignupClick}> Sign up </button>
        </div>
      </form>
    )
  }
}

Login.propTypes = {
  setLoading: PropTypes.func.isRequired,
  handleAuthError: PropTypes.func.isRequired
}
