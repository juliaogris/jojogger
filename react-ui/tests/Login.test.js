/* global it expect jest */
import React from 'react'
import ReactDOM from 'react-dom'
import Login from '../src/components/Login'
import renderer from 'react-test-renderer'
import ReactTestUtils from 'react-addons-test-utils'

jest.unmock('../src/util/api')
const api = require('../src/util/api')

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<Login
    setLoading={() => {}}
    handleAuthError={() => {}}
    />, div)
})

it('renders "Login" and matches snapshot', () => {
  const tree = renderer.create(<Login
    setLoading={() => {}}
    handleAuthError={() => {}}
    />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('calls RestAPI login on simulated Login Click', () => {
  const login = ReactTestUtils.renderIntoDocument(<Login
    setLoading={() => {}}
    handleAuthError={() => {}} />)
  login.setState({ email: 'j@x.co', password: 'pa$$word' })
  api.login = jest.fn(() => ({
    then: () => ({
      catch: () => {}
    })
  }))
  login.handleLoginClick({ preventDefault: () => {} })
  expect(api.login).toHaveBeenCalledWith('j@x.co', 'pa$$word')
})

it('calls RestAPI signup on simulated Signup Click', () => {
  const login = ReactTestUtils.renderIntoDocument(<Login
    setLoading={() => {}}
    handleAuthError={() => {}} />)
  login.setState({ email: 'j@x.co', password: 'pa$$word' })
  api.signup = jest.fn(() => ({
    then: () => ({
      catch: () => {}
    })
  }))
  login.handleSignupClick({ preventDefault: () => {} })
  expect(api.signup).toHaveBeenCalledWith('j@x.co', 'pa$$word')
})
