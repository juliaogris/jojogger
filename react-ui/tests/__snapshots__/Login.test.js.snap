exports[`test renders "Login" and matches snapshot 1`] = `
<form
  action={undefined}
  className="login">
  <input
    name="email"
    onChange={[Function]}
    placeholder="email@example.com"
    type="email"
    value="" />
  <input
    name="password"
    onChange={[Function]}
    placeholder="your password"
    type="password"
    value="" />
  <div
    className="button-row">
    <button
      onClick={[Function]}>
       Log in 
    </button>
    <button
      onClick={[Function]}>
       Sign up 
    </button>
  </div>
</form>
`;
